docs/docs/slot-validation-actions.mdx

```diff
@@ -0,0 +1,101 @@
+title: Slot Validation Actions
+abstract: A slot validation action is a special type of custom action, designed to handle custom extraction and/or validation of slot values. This can be used to validate slots with predefined mappings or extract slots with custom mappings.
+---
+
+Rasa SDK provides three base classes that you can use to inherit from in order to define your custom slot extraction and
+validation code:
+
+## `Action` class
+
+You can use this class to build a [custom action](./custom-actions.mdx) that is limited to returning only
+`SlotSet` and `BotUttered` events. Any other event type will be filtered out by the default action
+[`action_extract_slots`](./default-actions.mdx#action_extract_slots).
+The name of the custom action will have to be specified in the `action` key of the
+[custom slot mapping](./domain.mdx#custom-slot-mappings) in the domain.
+Note that the action name must be listed in the domain `actions` section too.
+
+## `ValidationAction` class
+
+You can extend the `ValidationAction` in the Rasa SDK to define custom extraction and / or validation of slots that
+can be set or updated outside of a form context.
+With this option, you do not need to specify the `action` key in the [custom slot mapping](./domain.mdx#custom-slot-mappings).
+
+Note that you do not need to implement the `name` method in the custom action extending `ValidationAction`, since this
+is already provided: `action_validate_slot_mappings`. However this name must be added to the domain `actions` list.
+
+:::note
+Slots mapped to a form in the mapping `conditions` will be ignored by `ValidationAction` unless you override
+the `required_slots` method to include the slots you want to extract and / or validate.
+:::
+
+### Validation of Slots with Predefined Mappings
+
+You must write functions named `validate_<slot_name>` for every predefined slot that you need validated.
+
+In the following example, the value for slot `location` is capitalized only if the extracted value is of type string:
+
+```python
+from typing import Text, Any, Dict
+
+from rasa_sdk import Tracker, ValidationAction
+from rasa_sdk.executor import CollectingDispatcher
+from rasa_sdk.types import DomainDict
+
+
+class ValidatePredefinedSlots(ValidationAction):
+    def validate_location(
+        self,
+        slot_value: Any,
+        dispatcher: CollectingDispatcher,
+        tracker: Tracker,
+        domain: DomainDict,
+    ) -> Dict[Text, Any]:
+        """Validate location value."""
+        if isinstance(slot_value, str):
+            # validation succeeded, capitalize the value of the "location" slot
+            return {"location": slot_value.capitalize()}
+        else:
+            # validation failed, set this slot to None
+            return {"location": None}
+```
+
+### Extraction of Custom Slot Mappings
+
+In order to define custom extraction code, you must write an `extract_<slot_name>` method for every slot with a
+custom slot mapping.
```

Comment:
What happens if the user specifies a custom action in the `action` key of a custom slot mapping, and then provides a validation method here? Does that work?

Reply:
Yup, this _should_ work, however there's a reported bug blocking this from working fully (by that I mean that the ValidationAction fails to run for the extracted slot value by the custom action from the `action` key), but should be fixed before the release 🤞🏼

Reply:
Got it, are we sure enough about that to update this section accordingly? Else that bug should get an additional "Change docs when you fix this" note

Reply:
I actually fixed the bug that I thought was causing the problem yesterday, but the problem persists so I created this issue in `rasa-sdk` repo: https://github.com/RasaHQ/rasa-sdk/issues/593
To be earnest, I don't know if this will be prioritised before the release - mainly because I think it should not be a recommended route of extraction and validation (I think it's better design for `actions.py` to have extraction and validation for one slot in one place, either the custom action or the `ValidationAction`; if the custom action is supposed to return a `SlotSet` event it should have the validated value already). I can add the note to change docs to its definition of done.